### PR TITLE
s/PatKind::Ident/PatKind::Binding/g

### PR DESCRIPTION
**Pull request by llogiq**
This failed the most recent build.

### CODE REVIEW

**Comment by Manishearth** [comment]
Still fails Travis. Also, version+changelog bump?

**Comment by llogiq** [comment]
First it has to work. Something changed beyond the name...

**Comment by mcarton** [comment]
You missed one.

**Comment by llogiq** [comment]
The copies lint appears to find more matches now, which is nice. I'm a bit stumped on the matches test, though. @mcarton any idea?

**Comment by mcarton** [comment]
This can warn:
```rust
    let _ = match Some(42) {
        Some(_) => 24,
        None => 24, //~ERROR this `match` has identical arm bodies
    };
```

This should not:
```rust
    let _ = match Some(42) {
        Some(foo) => 24,
        None => 24,
    };
```

Add this in the test and if it passes, r=me.

It also looks like you might need to rebase your local copy, there is still one `PatKind::Ident` left (see Travis).

**Comment by mcarton** [comment]
For the match test you might need `hir::PatKind::Path` instead (we’re matching `None` &co here).

File: clippy_lints/src/utils/hir.rs [hunk 1]
```diff
@@ -145,7 +145,7 @@ impl<'a, 'tcx: 'a> SpanlessEq<'a, 'tcx> {
             (&PatKind::TupleStruct(ref lp, ref la, ls), &PatKind::TupleStruct(ref rp, ref ra, rs)) => {
                 self.eq_path(lp, rp) && over(la, ra, |l, r| self.eq_pat(l, r)) && ls == rs
             }
-            (&PatKind::Ident(ref lb, ref li, ref lp), &PatKind::Ident(ref rb, ref ri, ref rp)) => {
```
Comment (mcarton) [inline clippy_lints/src/utils/hir.rs]: You might also want to add `PatKind::Path` here and add some tests for it. Something like:

```
if let None = Some(42) {
} else if let None = Some(42) {
}
```

See https://github.com/rust-lang/rust/pull/31685.

**Comment by llogiq** [comment]
Now on to matches...

File: clippy_lints/src/matches.rs [hunk 2]
```diff
@@ -200,7 +200,8 @@ fn check_single_match_opt_like(cx: &LateContext, ex: &Expr, arms: &[Arm], expr:
             }
             path.to_string()
         }
-        PatKind::Ident(BindByValue(MutImmutable), ident, None) => ident.node.to_string(),
+        PatKind::Binding(BindByValue(MutImmutable), ident, None) => ident.node.to_string(),
+        PatKind::Path(ref path) => path.to_string(),
```
Comment (mcarton) [inline clippy_lints/src/matches.rs]: Is `Binding` really necessary here? My understanding was that variant were now resolved (and we only care about some known paths).

**Comment by mcarton** [comment]
Looks like you failed your rebase/merge, some commits from master have appeared with a different hash.

File: clippy_lints/src/matches.rs (hunk 2 above)
Comment (llogiq) [inline clippy_lints/src/matches.rs]: I'm not sure, but I'm tired now and have to get up early tomorrow.

**Comment by llogiq** [comment]
I guess we can squash on merging. :wink:

**Pull request merged**